1단계 - 레거시 코드 리팩터링

sunggyupaik opened the pull request:
## 학습 관리 시스템(Learning Management System)

### 질문 삭제하기 요구사항
- 질문 데이터를 완전히 삭제하는 것이 아니라 데이터의 상태를 삭제 상태(deleted - boolean type)로 변경한다.
- 로그인 사용자와 질문한 사람이 같은 경우 삭제 가능하다.
- 답변이 없는 경우 삭제가 가능하다.
- 질문자와 답변글의 모든답변자가 같은경우 삭제가가능하다.
- 질문을 삭제할 때 답변 또한 삭제해야 하며, 답변의 삭제 또한 삭제 상태(deleted)를 변경
한다.
- 질문자와 답변자가 다른경우 답변을 삭제할 수 없다.
- 질문과 답변 삭제 이력에 대한 정보를 DeleteHistory를 활용해 남긴다.

### 리팩터링 요구사항
- nextstep.qna.service.QnaService의 deleteQuestion()는 앞의 질문 삭제 기능을 구현한 코드이다. 이 메소드는 단위 테스트하기 어려운 코드와 단위 테스트 가능한 코드가 섞여 있다.
- QnaService의 deleteQuestion() 메서드에 단위 테스트 가능한 코드(핵심 비지니스 로직)를 도메인 모델 객체에 구현한다.
- QnaService의 비지니스 로직을 도메인 모델로 이동하는 리팩터링을 진행할 때 TDD로 구현한다.
- QnaService의 deleteQuestion() 메서드에 대한 단위 테스트는 src/test/java 폴더 nextstep.qna.service.QnaServiceTest이다. 도메인 모델로 로직을 이동한 후에도 QnaServiceTest의 모든 테스트는 통과해야 한다.

sunggyupaik committed:
요구사항 정리

질문 삭제하기 요구사항과 리팩터링 요구사항을 정리합니다

sunggyupaik committed:
[refactor] 질문 삭제를 Question 도메인으로 이동하라

기존에 질문, 답변 삭제 상태 변경을 Service Layer에서 했습니다.
도메인에 책임을 질 수 있도록 Question 도메인으로 이동합니다

sunggyupaik committed:
[refactor] Question 답변 삭제 유효성을 Answer로 옮겨라

질문 작성자와 답변 작성자가 같은지 검사하는 로직을 Answer에게 넘겨
자신의 책임을 다할 수 있도록 개선합니다

sunggyupaik committed:
[refactor] 질문 삭제 서비스 메서드를 리팩토링 하라

질문을 조회하는 로직을 별도의 메소드로 분리합니다. 삭제된 질문과
답변들은 DB에 이력을 남기므로 DeleteHistory 리스트를 만들기
위해 메서드로 나누어 조합합니다

sunggyupaik committed:
[refactor] 삭제이력 생성을 개선하라

삭제이력 생성을 서비스가 아닌 도메인에서 하도록 개선합니다

sunggyupaik committed:
[refactor] 질문, 답변, 삭제 이력 각자의 책임을 부여하라

질문은 로그인 한 사용자가 삭제하는지, 답변은 로그인한 사용자가
삭제하는지 각자 책임을 지도록 개선합니다

삭제 이력은 질문과 답변을 변환하는 정적 팩터리 메서드를 추가합니다

sunggyupaik committed:
[refactor] 테스트에 사용히 필요없는 상수를 제거하라

Q1은 이번 테스트에서 필요하지 않는 상수이므로 제거합니다

sunggyupaik commented:
안녕하세요 step1 완료했습니다 
잘 부탁드립니다 :)

sunggyupaik commented:
이번 강의에서는 jdbc 쿼리를 짜는 것 같지 않아서 상세 쿼리는 짜지 않고 테스트에서 `@Mock`으로 대신하였습니다.

`QnaServieTest.java`에 2가지를 추가하였는데, `AnswerRepository`를 활용하여 `when(...)...`을 추가하였고 
`verifyDeleteHistories`를 `QnaService`에서 사용한 방식에 따라서 Arrays.asList 형태를 List로 변환하였습니다.

sunggyupaik committed:
[refactor] 생성시간, 수정시간을 클래스로 분리하라

생성시간, 수정시간을 별도의 클래스로 분리합니다

sunggyupaik committed:
[refactor] 질문의 답변을 일급 컬렉션으로 개선하라

질문 리스트를 일급 컬렉션으로 개선합니다

sunggyupaik commented:
`Quesiton`가 `Answers`를 변수로 가지고 있는데 서비스 로직에서 어떻게 표현해야 하는지 어려웠습니다.

2가지 방법을 생각했습니다.
- `Question`을 `Answer`로 `addAnswer()`를 통해 `Question`에 모두 합치고 작업
장점은 코드를 핵심 비지니스 변경을 예방 할 수 있으며 응집도가 있고 가독성이 좋습니다. 하지만 별도로 add작업을 해야 합니다.

- `Question` , `Answer` 따로 조회해서 각각 처리
여러 도메인의 작업이 명시되므로 어떤 작업을 하는지 빠르게 확인할 수 있습니다. 하지만 코드가 길어지면 가독성이 떨어지고 비지니스 코드가 변경에 취약합니다. 

후자로 한 이유는, `Question`과 `Answer` 모두 삭제 여부를 true로 변경해야 하므로 굳이 둘을 합치는 것보다 각각 처리하면 좋겠다고 생각했습니다.

삭제여부를 true로 고치고 수정 사항을 DB에 업데이트 할 때, 위의 고민에서 연장선으로 `repository.update(question)`에서 Question 과 Answer를 모두 삭제하도록 할지,  `repository.update(question)`, `repository.update(answer)` 처럼 처리해야 하는지도 고민이 되었습니다.

ohtaeg commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -24,26 +23,43 @@ public class QnAService {
+        questionRepository.update(questionId, question);
 
-        List<Answer> answers = question.getAnswers();
+        Answers answers = answersToBeDeleted(loginUser, questionId);
```
QnaService가 Answer의 존재를 알고있네요 :)
미션 요구사항에 나와있는 사진을 보면 다음과 같이 되어있는데요!

![image](https://github.com/next-step/java-lms/assets/20432323/62b28409-ea7b-4dc2-8d07-c2bddfb0e7f2)


https://github.com/next-step/java-lms/pull/262#issuecomment-1832205476


아무래도 고민해주셨던 부분으로인해 Answer와 Question을 따로 조회하도록 해주신 것 같네요 👍 👍 
요구사항을 배제하면 후자로 선택해주신 이유에 공감이 되네요 :)
저도 두가지 경우를 고려했을 것 같은데 저는 전자를 선택했을 가능성이 높아보이네요!

정답이 없는 부분이다보니 참고만해주세요 :)
Question과 Answer가 1:N으로 연관 관계를 맺고 있을 가능성이 높아
Answer에 대한 존재를 Question 외에 알리지 않는 목적으로 AnswerService를 별도로 만들거나
ORM을 통해 변경 감지를 이용하여 두 객체의 변경사항을 update하거나
말씀해주신 부분처럼 응집도 있게 처리했을 것 같네요 😄 

요구사항에 맞게 QnaServicer가 Answer의 존재를 모르도록 해주시면 좋을 것 같아요 :)
추가로 DeleteHistory의 생성도 QnaService가 아닌 다른 곳에서 해주시면 좋을 것 같습니다 😄

sunggyupaik replied:
답변 감사합니다, 사진에 나온 구조가 이번 과제에 방향이라 생각해서 로직 수정했습니다!

ohtaeg commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,9 +46,8 @@ public Long getId() {
-    public Answer setDeleted(boolean deleted) {
-        this.deleted = deleted;
-        return this;
+    private void setDeleted() {
```
setter의 접근제어를 최소화 👍

ohtaeg commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -15,13 +16,11 @@ public class Question {
 
     private NsUser writer;
 
-    private List<Answer> answers = new ArrayList<>();
+    private Answers answers = new Answers();
```
일급 컬렉션으로 👍

ohtaeg commented on src/main/java/nextstep/qna/domain/DeleteHistory.java:
```diff
@@ -26,6 +28,34 @@ public DeleteHistory(ContentType contentType, Long contentId, NsUser deletedBy,
         this.createdDate = createdDate;
     }
 
+    public static DeleteHistory from(Question question) {
+        return new DeleteHistory(
+                ContentType.QUESTION,
+                question.getId(),
+                question.getWriter(),
```
question.getId(), question.getWriter()와 같이 getter를 사용하여 외부에 정보를 노출하고 있는데요!
DeleteHistory를 만드는데 필요한 정보는 Question 에서 알고 있으므로,
Question 객체에게 없는 필요한 데이터를 주입해주면서 히스토리 생성을 명령하도록 한다면 어떨까요 :) ?

sunggyupaik replied:
네, getter를 최대한 사용하지 않는 방법으로 생각해보고 데이터구조 상 Question 객체에 없는 필요한 데이터를 주입하기 편한 구조이므로 구현 방식을 변경했습니다.

ohtaeg commented on src/main/java/nextstep/qna/domain/DeleteHistory.java:
```diff
@@ -26,6 +28,34 @@ public DeleteHistory(ContentType contentType, Long contentId, NsUser deletedBy,
+        return deleteHistories;
+    }
+
+    public static DeleteHistory from(Answer answer) {
+        return new DeleteHistory(
+                ContentType.ANSWER,
+                answer.getId(),
+                answer.getWriter(),
```
같은 피드백입니다 😄

ohtaeg commented on src/main/java/nextstep/qna/domain/DeleteHistory.java:
```diff
@@ -26,6 +28,34 @@ public DeleteHistory(ContentType contentType, Long contentId, NsUser deletedBy,
+                ContentType.ANSWER,
+                answer.getId(),
+                answer.getWriter(),
+                LocalDateTime.now()
```
LocalDateTime.now() 와 같은 데이터를 내부에서 생성할 경우 외부에서 변경이 어려워
추후 테스트 코드가 유연하지 않을 수 있다고 생각이 들기도 하는데요 :)
이런 데이터를 외부에서 주입받아보는 것은 어떨까요?
https://jojoldu.tistory.com/451

sunggyupaik replied:
좋은 레퍼런스 감사합니다 :)
어디서부터 주입을 해야할지 고민을 했는데, 이왕 할거면 첫 시작부터 주입을 할 수 있어야 내가 테스트에서도 통제가 가능하다고 생각하여 서비스 레이어에서부터 주입이 가능하도록 개선했습니다.  `deleteQuestion(NsUser loginUser, long questionId, LocalDateTime date)`

ohtaeg reviewed (changes requested):
안녕하세요 성규님!
마지막 미션을 같이 함께 할 리뷰어 오태경이라고 합니다 👋 
잘 부탁드리며, 리뷰가 늦어서 죄송합니다 🙏 
1단계를 잘해주셨네요 👍 요구사항에 맞게 좀 더 리팩토링을 하면 좋을 것 같다 Request changes 하였습니다!! 피드백 확인 후 재요청 주시면 감사하겠습니다 :)

sunggyupaik committed:
[refactor] 답변을 질문 도메인에서 관리하라

질문 서비스 레이어에서 답변과 질문 도메인을 나눠서 관리하고 있지만
연관관계를 고려하여 질문 도메인에서 관리하도록 개선합니다

sunggyupaik committed:
[refactor] 삭제 이력의 의존을 변경하라

기존에는 삭제 이력이 질문과 답변을 의존했습니다. 하지만 질문과 답변의
도메인 정보를 get으로 꺼내했으므로 반대의 의존성을 만들어 노출을
최소화합니다

sunggyupaik committed:
[refactor] 질문 도메인에서 삭제 이력을 관리하라

질문 도메인이 삭제이력을 관리할 수 있도록 합니다.
따라서 도메인에서 질문과 답변을 삭제이력으로 만드는 작업을 합니다

sunggyupaik committed:
[feat] 삭제이력 일급 컬렉션을 추가하라

삭제이력을 리스트로 관리하지 않고 일급 컬렉션으로 관리합니다
삭제 이력에 질문과 답변을 추가하는 행위와 삭제이력 상태를 관리합니다

sunggyupaik commented:
> 저도 두가지 경우를 고려했을 것 같은데 저는 전자를 선택했을 가능성이 높아보이네요!

Question과 Answer 부분은 저도 jpa 관련되었다면 1:N 관계를 생각하고 연관관계를 통한 한방 조회를 생각했었는데요
이번 과제에서는 jdbc로 구현 되었다보니 `questionRepository.findById(questionId).orElseThrow(NotFoundException::new);`  에서 한번에 Question과 Answer가 조회가 가능할까? Question과 Answer를 각각조회하고 Question에 합쳐서 작업해야 하나? 생각하다보니 차라리 별도로 해보자 생각했었습니다.

sunggyupaik commented:
Question과 Answer의 `delete()` 메서드는 바로 DeleteHistory를 리턴하는 것보다, 다양한 객체의 쓰임새가 활용될 수 있다고 생각이 들어 자신의 객체를 리턴하도록 하였습니다. 대신에 `toDeleteHistories` 메서드에서 Question과 Answer를 조립하여 리턴하였습니다.

ohtaeg commented:
> Question과 Answer 부분은 저도 jpa 관련되었다면 1:N 관계를 생각하고 연관관계를 통한 한방 조회를 생각했었는데요
이번 과제에서는 jdbc로 구현 되었다보니 questionRepository.findById(questionId).orElseThrow(NotFoundException::new); 에서 한번에 Question과 Answer가 조회가 가능할까? Question과 Answer를 각각조회하고 Question에 합쳐서 작업해야 하나? 생각하다보니 차라리 별도로 해보자 생각했었습니다.

저도 다시 생각해보니 jdbc인 경우 를 고려해보았을 때 각각 조회할 것 같네요!! 👍 👍

ohtaeg commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -43,13 +42,21 @@ public Answer(Long id, NsUser writer, Question question, String contents) {
         this.contents = contents;
     }
 
+    public DeleteHistory toDeleteHistory(LocalDateTime time) {
```
> toDeleteHistories 메서드에서 Question과 Answer를 조립하여 리턴하였습니다.

👍 👍

ohtaeg commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -24,26 +23,16 @@ public class QnAService {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
+    public void deleteQuestion(NsUser loginUser, long questionId, LocalDateTime date) {
```
👍 👍

ohtaeg commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,46 @@
+        this.answers = answers;
+    }
+
+    public Answers delete(NsUser loginUser) {
+        List<Answer> deletedAnswers = new ArrayList<>();
+        for (Answer answer : answers) {
+            deletedAnswers.add(answer.delete(loginUser));
+        }
+
+        return new Answers(deletedAnswers);
+    }
+
+    public void add(Answer answer) {
+        this.answers.add(answer);
+    }
+
+    public DeleteHistories toDeleteHistories(LocalDateTime time) {
```
delete()와 toDeleteHistories()를 합치는 것도 괜찮아보이네요 😄 
(수정을 위한 리뷰는 아니오니 참고만 해주세요!!)
```suggestion
    public DeleteHistories delete(NsUser user, LocalDateTime time) {
```

sunggyupaik replied:
네, 해당 부분은 인지하고 넘어가겠습니다! 감사합니다

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨네요 👍 👍 
다음 미션도 화이팅입니다 🔥
